source_test.go

```diff
@@ -88,3 +88,66 @@ func TestRead_StructuredData(t *testing.T) {
 	err = json.Unmarshal(bytes, &recStruct) //nolint:staticcheck // test struct
 	is.NoErr(err)
 }
+
+func TestSource_Read_SleepGenerate(t *testing.T) {
+	is := is.New(t)
+	cfg := map[string]string{
+		Fields:       "id:int",
+		Format:       FormatRaw,
+		ReadTime:     "10ms",
+		SleepTime:    "200ms",
+		GenerateTime: "50ms",
+	}
+
+	underTest := NewSource()
+	t.Cleanup(func() {
+		_ = underTest.Teardown(context.Background())
+	})
+
+	err := underTest.Configure(context.Background(), cfg)
+	is.NoErr(err)
+
+	type result struct {
+		err      error
+		duration time.Duration
+	}
+
+	// first read: sleep time + read time + little bit of buffer
+	results := make(chan result)
+	go func() {
+		start := time.Now()
+		_, err := underTest.Read(context.Background())
+
+		results <- result{err: err, duration: time.Since(start)}
+	}()
+
+	select {
+	case r := <-results:
+		is.NoErr(r.err)
+		is.True(r.duration >= 210*time.Millisecond) // expected source to sleep for given time
+	case <-time.After(220 * time.Millisecond):
+		is.Fail() // timed out waiting for record
+	}
+
+	// we have 40ms left for generating new records
+	// (50ms total, minus the 10ms for the first record)
+	// so we read 4 more records here
+	results = make(chan result)
+	go func() {
+		start := time.Now()
+		for i := 1; i <= 4; i++ {
+			_, err = underTest.Read(context.Background())
+			is.NoErr(err)
+		}
+
+		results <- result{err: err, duration: time.Since(start)}
+	}()
+
+	select {
+	case r := <-results:
+		is.NoErr(r.err)
+		is.True(r.duration >= 40*time.Millisecond) // expected source to sleep for given time
+	case <-time.After(50 * time.Millisecond):
+		is.Fail() // timed out waiting for record
+	}
```

Comment:
Can we add another read after this select? That one should then again happen after ~210ms, right?

Reply:
Makes sense, I just pushed it.